### ISSUE TITLE
Dictionary.items() getter function

### ISSUE THREAD

**Issue by VincentRagusa** [godotengine/godot-proposals#1965]
### Describe the project you are working on
I'm using godot to build a sort of "toy" (as opposed to a game) where you can simulate evolution of little critters and interact with them from the UI.

### Describe the problem or limitation you are having in your project
I'm using a dictionary to track information based on IDs the entries are of the form key:[value1, [values2...],[values3...]].
My problem is that I would like to display the contents of the dictionary on the screen sorted by the average of [values2]. My plan was to do something similar to this python code: "dictionary.sort(key=lambda x: average(x[1]) )". i need to be able to convert the dictionary into an array of key/value pairs. something like [(key_0, [value1_0, [values2_0...],[values3_0...]),...,(key_n, [value1_n, [values2_n...],[values3_n...])]

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Basically I think we should just have a .items() function on dictionaries like python has. (don't be mad)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Assuming the dictionary is a hash table under the hood you would have to basically implement the thing im doing manually, iterate over the keys, extract the values and match them up correctly before returning them as a list of pairs. Maybe there's a better more efficient way that can leverage the underlying data structure.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
The closest method i can imagine with the tools we already have is to call  .keys() iterate over them using a for loop, make the list of key/value pairs, and then sort that list of pairs using a custom sorting function because we also don't have lambdas (The custom sort can already be handled using this method  via https://godotengine.org/qa/32053/how-to-sort-array). But this is extremely messy and id say 2 helper functions is a lot of extra code to just sort a dictionary by its values.

### Is there a reason why this should be core and not an add-on in the asset library?
dictionary is a core object, all of its functions should be included with it. extracting key/value pairs seems crazy as an external library function

**Comment by Calinou** [comment]
Have you looked at `Dictionary.values()`?

**Comment by VincentRagusa** [comment]
> Have you looked at `Dictionary.values()` which is the value counterpart to `Dictionary.keys()`?

Yes, but i believe that only gives you the values, not a key/value pair. There is a risk in just calling keys() and values() and then iterating over them at the same time because dictionaries are unordered and there is not usually a guarantee that the two calls will result in the same ordering of items. This might be solved in godot 3.x since dictionaries are actually ordered by the insertion order iirc. However there is still the issue of pairing the items together and sorting them. with no built in function like python's zip() to quickly combine two arrays into an array of pairs a custom zip function must be written by the user.
So, while .values() may provide easier access to the values than iterating over the keys, any attempt to use it will rely on the assumption that the items returned by keys() and values() correspond to the same ordering, and the user still has to write a custom zip() function.

**Comment by Calinou** [comment]
> There is a risk in just calling keys() and values() and then iterating over them at the same time because dictionaries are unordered and there is not usually a guarantee that the two calls will result in the same ordering of items

The Dictionary class *does* guarantee ordering of elements since Godot 3.0.

**Comment by KoBeWi** [comment]
> The Dictionary class does guarantee element ordering by insertion since Godot 3.0

Unless they are serialized, in which case they are ordered alphabetically.

**Comment by bluenote10** [comment]
> Have you looked at Dictionary.values() which is the value counterpart to Dictionary.keys()?

That doesn't really solve the use case of conveniently iterating keys and values, does it?

Also there should be a possibility of a performance improvement when the loop can lookup keys and values simultaneously. 

**Current solution**
```gdscript
var data = {}

for key in data: # keys get extracted here
    # key -> value lookup happens somewhere in loop body
    # potentially even repeatedly if the user doesn't pull out the value into a local variable
    data[key]
```

**More convenient + possibly more efficient:**
```gdscript
var data = {}

for key, value in data.items(): # keys and values get extracted here
    # both already available in loop body
```

The question is what `.items()` would return exactly. If it has to build a new array of tuples in memory, it is not much of a gain. Ideally it should return some kind of iterator type that can be passed around. Not sure how easy it would be to accomplish that.

Perhaps a similar solution as in Nim would work: In Nim the context of the for loop can lead to applying a different kind of iterator to the expression given in the loop. Basically if the compiler sees `for key, value in ...` it does something different then when the context is `for single_iterable in ...`. As a result the three major dict iteration variants look like this in Nim (called `tables` there other than that pretty close to GDScript):

```nim
import tables

var table = initTable[int, string]()
table[10] = "a"
table[20] = "b"

for key in table.keys():
  echo key

for value in table.values():
  echo value

for key, value in table:
  echo key, value
```

<details>
<summary>outputs</summary>

```
20
10
b
a
20b
10a
```
</details>

This would basically mean that `.items()` isn't necessarily something that can be called manually or passed around. But perhaps special-casing `for key, value in ...` for dictionaries would already cover the most important use case of iterating keys and values simultaneously?
